settings: Do not use bootstrap tabs in bots panel.

sahil839 opened the pull request:
This PR replaces the boostrap tabs with our own tabs implementation in components.ts.
This is part of the bootstrap removal project.

<!-- Describe your pull request here.-->

<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![Peek 2024-01-11 23-58](https://github.com/zulip/zulip/assets/35494118/17d2e2ef-9220-4b56-8d9c-2f2ff34e6245)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on web/tests/settings_bots.test.js:
```diff
@@ -102,45 +92,6 @@ test("generate_botserverrc_content", () => {
-
-    assert.ok(!forms.$active.visible());
-    assert.ok(forms.$inactive.visible());
-});
```
Removed this test, since it would have needed mocked functions and elements for code in `components.toggle` and we already test the code in `components.ts`.

sahil839 commented:
@alya I think you can have a first look at this to see if we need any UI changes. For now I have just changed the tabs and nothing else.

timabbott commented:
Looks very plausible to me.

alya commented:
Hm, I'm finding it a bit hard to parse the tabs as controlling the area below. Maybe putting a line above the tabs would help?

alya commented:
While we're here, would it be easy to move this line to the Active bots tab?

> Download config of all active outgoing webhook bots in Zulip Botserver format. 

It relates to active bots, so structurally it belongs there.

sahil839 commented:
> Hm, I'm finding it a bit hard to parse the tabs as controlling the area below. Maybe putting a line above the tabs would help?

What do you mean by line here? Do we need a blank line, i.e. a linebreak or a thin horizontal divider?

sahil839 commented:
> While we're here, would it be easy to move this line to the Active bots tab?
> 
> > Download config of all active outgoing webhook bots in Zulip Botserver format.
> 
> It relates to active bots, so structurally it belongs there.

Yeah, can do that. Do we want to show that text if the user has no active bots?

timabbott commented:
I think a thin horizontal divider was what she had in mind. 

And yeah, we probably don't need that text if there are no active bots.

alya commented:
> I think a thin horizontal divider was what she had in mind.

Heh, yep!

> And yeah, we probably don't need that text if there are no active bots.

Makes sense to me.

sahil839 commented on web/templates/settings/bot_settings.hbs:
```diff
@@ -13,22 +13,24 @@
+        </div>
 
-        <ol id="inactive_bots_list" class="bots_list" data-bot-settings-section="inactive-bots" data-empty="{{t 'You have no inactive bots.' }}">
+        <ol id="inactive_bots_list" class="bots_list bots_section" data-bot-settings-section="inactive-bots" data-empty="{{t 'You have no inactive bots.' }}">
```
We could have added a containing `div` for this list as well but it is necessary, but still might help in make the code especially the classnames and CSS easy to read and avoid bugs.

timabbott replied:
Yeah I would add `inactive_bots_list_container` just to make the code/components more parallel.

sahil839 commented:
Updated the PR as per feedback. I am not sure about the design of divider, so that may still need some tweaks.

![Screenshot from 2024-01-19 00-34-01](https://github.com/zulip/zulip/assets/35494118/cb0bbf0f-0c4a-4b0f-98e0-fcbd0d7c36a8)

alya commented:
> Download config of all active outgoing webhook bots in Zulip Botserver format. 

I guess we can actually further restrict when this is shown to situations where there's at least one outgoing webhook bot? Not worth worrying about if it's non-trivial, though.

alya commented:
The divider line looks OK to me, but the vertical spacing feels too tight. Can we add a good amount more space above? Possibly a bit more space below would look better as well.

alya commented:
While looking around the settings to see if there are dividers elsewhere, I noticed an extraneous divider here. Could you please add a commit to drop it?
![Screenshot 2024-01-18 at 4 23 37 PM](https://github.com/zulip/zulip/assets/2090066/ca2daacb-defa-4c96-b29c-9873560e0ee3)

sahil839 commented:
Updated the PR -
- The text is now shown only if there is atleast one active outgoing webhook bot.
- Increased space below and above the header.
- Added commit to remove the extraneous separator.

![image](https://github.com/zulip/zulip/assets/35494118/ade692b1-4ef7-4e45-b7c2-98e62c362668)

timabbott commented on web/src/settings_bots.js:
```diff
@@ -24,6 +24,7 @@ import * as user_profile from "./user_profile";
 const INCOMING_WEBHOOK_BOT_TYPE = 2;
 const OUTGOING_WEBHOOK_BOT_TYPE = "3";
+const OUTGOING_WEBHOOK_BOT_TYPE_INT = 3;
 const EMBEDDED_BOT_TYPE = "4";
```
Should these just all be in integer format? I see `INCOMING_WEBHOOK_BOT_TYPE` is already 2. Maybe worth having a comment about what's going on here or just cleaning it up.

sahil839 replied:
The string values are used to compare the value selected in "Bot type" dropdown because `$("#create_bot_type").val()` returns string. We can also do `OUTGOING_WEBHOOK_BOT_TYPE.toString()` while comparing.

timabbott replied:
That seems probably nicer. Not a blocker for this PR.

timabbott commented:
Merged, thanks @sahil839!
